[JSC] Exnref should be subtype of anyref

Constellation opened the pull request:
#### 149c01577b7eeb5c5fc619b992d292e8eb6027ae
<pre>
[JSC] Exnref should be subtype of anyref
<a href="https://bugs.webkit.org/show_bug.cgi?id=285783">https://bugs.webkit.org/show_bug.cgi?id=285783</a>
<a href="https://rdar.apple.com/142474856">rdar://142474856</a>

Reviewed by Yijia Huang.

exnref should be subtype of anyref. So WasmTable should be OK for
exnref when it is externref table.

* JSTests/wasm/stress/exnref-table.js: Added.
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isExnref):
(JSC::Wasm::isSubtype):

Canonical link: <a href="https://commits.webkit.org/288811@main">https://commits.webkit.org/288811@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9651a78dde4e4b08678bc2feb94f0adce0f1198d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89572 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35501 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4200 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12100 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/65706 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23552 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87536 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/76766 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46001 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30997 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34549 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77454 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31772 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90953 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83507 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11757 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11984 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72590 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17700 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3192 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13082 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11709 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17185 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105926 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11558 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25556 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15034 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13331 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on JSTests/wasm/stress/exnref-table.js:
```diff
@@ -0,0 +1,4 @@
+var data = [ 0x0, 0x61, 0x73, 0x6d, 0x1, 0x0, 0x0, 0x0, 0x1, 0x99, 0x0, 0x3, 0x60, 0x3, 0x7f, 0x7f, 0x7f, 0x1, 0x7f, 0x60, 0xb, 0x7e, 0x7f, 0x7f, 0x7f, 0x7c, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x0, 0x60, 0x0, 0x0, 0x4, 0x8d, 0x0, 0x3, 0x69, 0x1, 0x1, 0x35, 0x6a, 0x1, 0x17, 0x1c, 0x6b, 0x1, 0x15, 0x41 ];
```
It seems that the `data` corresponds to the following WAT:

```wat
(module
  (type (;0;) (func (param i32 i32 i32) (result i32)))
  (type (;1;) (func (param i64 i32 i32 i32 f64 i32 i32 i32 i32 i32 i32)))
  (type (;2;) (func))
  (table (;0;) 1 53 exnref)
  (table (;1;) 23 28 arrayref)
  (table (;2;) 21 65 structref)
)
```

However, for this PR's test, I believe the `func` definitions and the `arrayref` and `structref` tables are unnecessary.

Shouldn't the WAT be shortened to something like the following?

```js
// (module
//   (table (;0;) 1 53 exnref)
// )
var data = [0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, 0x04, 0x05, 0x01, 0x69, 0x01, 0x01, 0x35];
```

At least, the corresponding WAT should be included as a comment for better readability.

Constellation replied:
This is coming from fuzzer bug and not particularly focusing on reducing the test case since it is already small, so I'll leave it as is.

kmiller68 replied:
> This is coming from fuzzer bug and not particularly focusing on reducing the test case since it is already small, so I'll leave it as is.

Can we at least include the current WAT as a comment? It's pretty annoying to reconstruct it from the binary when you need it.

Constellation replied:
Done.

sosukesuzuki reviewed (approved):
(This is informal approve)

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Exnref should be subtype of anyref
https://bugs.webkit.org/show_bug.cgi?id=285783
rdar://142474856

Reviewed by Yijia Huang.

exnref should be subtype of anyref. So WasmTable should be OK for
exnref when it is externref table.

* JSTests/wasm/stress/exnref-table.js: Added.
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isExnref):
(JSC::Wasm::isSubtype):

Canonical link: https://commits.webkit.org/288811@main
